### PR TITLE
Fix key resolution text #2014 and more

### PR DESCRIPTION
**Pull request by robander**
The DITA 1.3 spec topic about [generated text based on key references](http://docs.oasis-open.org/dita/dita/v1.3/errata01/os/complete/part1-base/archSpec/base/processing-keyref-for-text.html#processing_key_references) is clear that when the primary rules do not work, all elements with `@keyref` should look to `<linktext>`. This is not working today for elements without `@href`, as reported in #2014.

After starting with @raducoravu 's suggested fix, I extended my test case and found quite a few more edge cases that are either broken (wrong output) or bad (no output when we could provide something). This pull request covers all of the following updates:

- In rule number 2, the spec topic says non-linking elements should look for `<keyword>` inside of `<keywords>` and consider that as matching content. **Today: broken for all elements when keyword appears outside of keywords.** The code looks for `<keyword>` _anywhere_ within the key definition. So if I have `<linktext><keyword>book</keyword> title</linktext>`, matching text for all elements becomes "book" instead of "book title".
- The spec explicitly says that for linking elements, if there are multiple keyword elements in `<keywords>`, all should be used (rule 3 has an example). **Today: broken, we only use the first.** This pull request takes them all, though I consider this an edge case - the rule carried over from DITA 1.2.
- Getting to rule 4, this request adds support for the full content of `<linktext>` as the next fallback for all elements. **Today: broken for all non-linking elements, broken for all elements if keyword appears in linktext.**
- Based on rule 5 in the spec topic, linking elements today fall back to `@navtitle`. **Today: broken for navtitle element, sort of working for navtitle attribute on linking elements, and gives null result for non-linking elements.** There are a couple of problems here:
  - If using navtitle, it should prefer `<navtitle>` (today that is ignored), so I've added that.
  - Non linking elements never get here, and give bad output. Having reached the end of the specification algorithm, with no good output, I think it's foolish to ignore the available `navtitle`, so I've added that as another fallback for non-linking elements.
  - If the target is a local DITA topic, we should _not_ pull the `navtitle` unless that title is locked, because the true title will be pulled from that topic later
- Additionally in rule 5, the specification says that "normal rules apply", which is clearly implementation specific. In links without keys, we fall back to the URI as the last resort, so I've added that as well. Again this is only for non-DITA or non-local content, because local DITA should wait and retrieve link text from the target.

Basically, the focus of this report is on
1. Getting linktext to work so that key definitions have a universal fallback (this is specifically described in the spec as a best practice, and is broken today.
1. Fixing output that today is broken (or just bad), and can easily be made correct (or at least reasonable).

### CODE REVIEW

**Comment by robander** [comment]
Looks like this also addresses #1590.

File: src/main/java/org/dita/dost/writer/KeyrefPaser.java [hunk 1]
```diff
@@ -249,15 +249,23 @@ public void endElement(final String uri, final String localName, final String na
                     // grab keyword or term from key definition
                     if (!hasSubElem.peek() && currentElement != null) {
                         NodeList nodeList = elem.getElementsByTagName(TOPIC_KEYWORD.localName);
+                        final List<Element> keywordsInKeywords = new ArrayList<>();
+                        for (int index = 0; index < nodeList.getLength(); index++) {
+                            final Element keywordParent = (Element) nodeList.item(index).getParentNode();
+                            final DitaClass parentcls = keywordParent.getAttribute(ATTRIBUTE_NAME_CLASS) != null ? new DitaClass(keywordParent.getAttribute(ATTRIBUTE_NAME_CLASS)) : null;
+                            if (parentcls != null && parentcls.equals(TOPIC_KEYWORDS)) {
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/writer/KeyrefPaser.java]: This should use `matches`, not `equals`.

**Comment by jelovirt** [comment]
Converted for-loop with index to stream API, intent easier to read.

Extracted `writeLinktext` methods to separate methods, also for readability.
